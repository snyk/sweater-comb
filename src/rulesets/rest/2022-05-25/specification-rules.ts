import { RuleError, Ruleset, SpecificationRule } from "@useoptic/rulesets-base";
import { dotCase, pascalCase, snakeCase } from "change-case";
import { links } from "../../../docs";
import { stabilityKey } from "./constants";

interface componentName {
  localName: string;
  localProp?: string;
  ns?: string;
}

/**
 * Decode the component name from a possibly namespace and property-qualified OpenAPI component name.
 *
 * @param name An OpenAPI component name
 * @returns Tuple containing the local name (without namespace) and its namespace prefix (undefined if not namespaced).
 */
const decodeComponentName = (name: string): componentName => {
  const parts = name.split(".");
  if (parts.length > 1) {
    const lastPart = parts[parts.length - 1];
    if (lastPart === snakeCase(lastPart)) {
      // Component is derived from a Typespec model property, of the form namespace.ModelName.property_name.
      // This is the case for shared parameter or header components, for example.
      return {
        localName: parts[parts.length - 2],
        localProp: lastPart,
        ns: parts.slice(0, parts.length - 2).join("."),
      };
    } else {
      // Component is derived from a Typespec model, of the form namespace.ModelName.
      // This is the case for schema components.
      return {
        localName: parts[parts.length - 1],
        ns: parts.slice(0, parts.length - 1).join("."),
      };
    }
  } else {
    return { localName: name };
  }
};

const componentNameCase = new SpecificationRule({
  name: "component names",
  docsLink: links.standards.componentNaming,
  rule: (specificationAssertions) => {
    specificationAssertions.requirement(
      "use pascal case for component names",
      (specification) => {
        const componentTypes = Object.keys(specification.raw.components || {});
        for (const componentType of componentTypes) {
          if (componentType.startsWith("x-")) {
            continue;
          }
          if (componentType === "securitySchemes") {
            continue;
          }
          const componentNames = Object.keys(
            specification.raw.components?.[componentType] || {},
          );
          for (const componentName of componentNames) {
            const { localName, localProp, ns } =
              decodeComponentName(componentName);
            if (pascalCase(localName) !== localName) {
              throw new RuleError({
                message: `Expected ${localName} to be pascal case in component ${componentName}`,
              });
            }
            if (localProp && snakeCase(localProp) !== localProp) {
              throw new RuleError({
                message: `Expected ${localProp} to be snake case in component ${componentName}`,
              });
            }
            if (ns && dotCase(ns) !== ns) {
              throw new RuleError({
                message: `Expected ${ns} to be dot case in component ${componentName}`,
              });
            }
          }
        }
      },
    );
  },
});

const listOpenApiVersions = new SpecificationRule({
  name: "list open api version",
  // Only applicable to compiled OAS; resource versions do not need to declare this
  matches: (specification) => specification.value[stabilityKey] === undefined,
  docsLink: links.standards.openApiVersions,
  rule: (specificationAssertions) => {
    specificationAssertions.requirement(
      "list the available versioned OpenAPI specifications",
      (specification) => {
        const pathUrls = Object.keys(specification.raw.paths);
        if (!pathUrls.includes("/openapi")) {
          throw new RuleError({
            message: "Expected route /openapi to be included",
          });
        }
      },
    );
  },
});

const getOpenApiVersions = new SpecificationRule({
  name: "get open api versions",
  // Only applicable to compiled OAS; resource versions do not need to declare this
  matches: (specification) => specification.value[stabilityKey] === undefined,
  docsLink: links.standards.openApiVersions,
  rule: (specificationAssertions) => {
    specificationAssertions.requirement(
      "provide versioned OpenAPI specifications",
      (specification) => {
        const pathUrls = Object.keys(specification.raw.paths);
        if (!pathUrls.includes("/openapi/{version}")) {
          throw new RuleError({
            message: "Expected route /openapi/{version} to be included",
          });
        }
      },
    );
  },
});

const tags = new SpecificationRule({
  name: "open api version names",
  docsLink: links.standards.tags,
  rule: (specificationAssertions) => {
    specificationAssertions.requirement(
      "have name and description for tags",
      (specification) => {
        const tags = specification.value.tags || [];
        for (const tag of tags) {
          if (!("name" in tag)) {
            throw new RuleError({
              message: "name is not in tag",
            });
          }
          if (!("description" in tag)) {
            throw new RuleError({
              message: "description is not in tag",
            });
          }
        }
      },
    );
  },
});

const noDiscriminatorMapping = new SpecificationRule({
  name: "no mapping objects in discriminators",
  docsLink: links.standards.polymorphicObjects,
  rule: (specificationAssertions) => {
    specificationAssertions.addedOrChanged(
      "no mapping objects in discriminators",
      (specification) => {
        // propertyName is the only required field on a discriminator object,
        // plus we only care about the objects with mappings
        const discriminators = findObjectsWithFields([
          "propertyName",
          "mapping",
        ])(specification.raw);

        if (discriminators.length != 0) {
          throw new RuleError({
            message: "Mapping object is not permitted in discriminators",
          });
        }
      },
    );
  },
});

const findObjectsWithFields = (fields: string[]) => (from: unknown) => {
  if (typeof from !== "object" || from == null) {
    return [];
  }
  return Object.values(from).flatMap((v) => {
    if (typeof v !== "object" || v == null) {
      return [];
    }
    if (Array.isArray(v)) {
      return v.flatMap(findObjectsWithFields(fields));
    }
    if (fields.every((field) => field in v)) {
      return [v];
    }
    return findObjectsWithFields(fields)(v);
  });
};

export const specificationRules = new Ruleset({
  name: "specification rules",
  rules: [
    componentNameCase,
    tags,
    getOpenApiVersions,
    listOpenApiVersions,
    noDiscriminatorMapping,
  ],
});
