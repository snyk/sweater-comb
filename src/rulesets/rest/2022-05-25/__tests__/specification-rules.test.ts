import { OpenAPIV3 } from "@useoptic/openapi-utilities";
import { RuleRunner, TestHelpers } from "@useoptic/rulesets-base";
import { context } from "./fixtures";

import { specificationRules } from "../specification-rules";
import { stabilityKey } from "../constants";

const baseJson = TestHelpers.createEmptySpec();

describe("specification rules", () => {
  test("fails when components are not PascalCase", () => {
    const afterJson = {
      ...baseJson,
      [stabilityKey]: "wip",
      paths: {},
      components: {
        parameters: {
          OrgId: {
            name: "orgId",
            in: "path",
          },
          ThingId: {
            name: "thingId",
            in: "path",
          },
        },
        schemas: {
          thingResourceResponse: {
            type: "object",
            description: "Response containing a single thing resource object",
            properties: {},
          },
        },
      },
    } as OpenAPIV3.Document;
    const ruleRunner = new RuleRunner([specificationRules]);
    const ruleInputs = {
      ...TestHelpers.createRuleInputs(baseJson, afterJson),
      context,
    };
    const results = ruleRunner.runRulesWithFacts(ruleInputs);

    expect(results.length).toBeGreaterThan(0);
    expect(results.every((result) => result.passed)).toBe(false);
    expect(results).toMatchSnapshot();
  });

  describe("compiled specs", () => {
    test("fails when /openapi endpoint isn't specified", () => {
      const afterJson: OpenAPIV3.Document = {
        ...baseJson,
        paths: {
          "/openapi": {},
        },
      };
      const ruleRunner = new RuleRunner([specificationRules]);
      const ruleInputs = {
        ...TestHelpers.createRuleInputs(baseJson, afterJson),
        context,
      };
      const results = ruleRunner.runRulesWithFacts(ruleInputs);

      expect(results.length).toBeGreaterThan(0);
      expect(results.every((result) => result.passed)).toBe(false);
      expect(results).toMatchSnapshot();
    });

    test("fails when /openapi/{versions} endpoint isn't specified", () => {
      const afterJson: OpenAPIV3.Document = {
        ...baseJson,
        paths: {
          "/openapi/{versions}": {},
        },
      };
      const ruleRunner = new RuleRunner([specificationRules]);
      const ruleInputs = {
        ...TestHelpers.createRuleInputs(baseJson, afterJson),
        context,
      };
      const results = ruleRunner.runRulesWithFacts(ruleInputs);

      expect(results.length).toBeGreaterThan(0);
      expect(results.every((result) => result.passed)).toBe(false);
      expect(results).toMatchSnapshot();
    });
  });

  describe("uncompiled specs", () => {
    test("passes when /openapi endpoint isn't specified", () => {
      const afterJson = {
        ...baseJson,
        [stabilityKey]: "wip",
        paths: {
          "/openapi/{versions}": {},
        },
      } as OpenAPIV3.Document;
      const ruleRunner = new RuleRunner([specificationRules]);
      const ruleInputs = {
        ...TestHelpers.createRuleInputs(baseJson, afterJson),
        context,
      };
      const results = ruleRunner.runRulesWithFacts(ruleInputs);

      expect(results.length).toBeGreaterThan(0);
      expect(results.every((result) => result.passed)).toBe(true);
      expect(results).toMatchSnapshot();
    });

    test("passes when /openapi/{versions} endpoint isn't specified", () => {
      const afterJson = {
        ...baseJson,
        [stabilityKey]: "wip",
        paths: {
          "/openapi": {},
        },
      } as OpenAPIV3.Document;
      const ruleRunner = new RuleRunner([specificationRules]);
      const ruleInputs = {
        ...TestHelpers.createRuleInputs(baseJson, afterJson),
        context,
      };
      const results = ruleRunner.runRulesWithFacts(ruleInputs);

      expect(results.length).toBeGreaterThan(0);
      expect(results.every((result) => result.passed)).toBe(true);
      expect(results).toMatchSnapshot();
    });
  });
});
