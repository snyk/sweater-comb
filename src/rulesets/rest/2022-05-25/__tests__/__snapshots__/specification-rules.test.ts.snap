// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`specification rules compiled specs fails when /openapi endpoint isn't specified 1`] = `
[
  {
    "change": {
      "location": {
        "conceptualLocation": {},
        "conceptualPath": [],
        "jsonPath": "",
        "kind": "specification",
      },
      "value": {
        "info": {
          "title": "Empty",
          "version": "0.0.0",
        },
        "openapi": "3.1.0",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/standards/rest.md#referenced-entities",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "component names",
    "passed": true,
    "received": undefined,
    "type": "requirement",
    "where": "specification",
  },
  {
    "change": {
      "location": {
        "conceptualLocation": {},
        "conceptualPath": [],
        "jsonPath": "",
        "kind": "specification",
      },
      "value": {
        "info": {
          "title": "Empty",
          "version": "0.0.0",
        },
        "openapi": "3.1.0",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/standards/rest.md#tags",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "open api version names",
    "passed": true,
    "received": undefined,
    "type": "requirement",
    "where": "specification",
  },
  {
    "change": {
      "location": {
        "conceptualLocation": {},
        "conceptualPath": [],
        "jsonPath": "",
        "kind": "specification",
      },
      "value": {
        "info": {
          "title": "Empty",
          "version": "0.0.0",
        },
        "openapi": "3.1.0",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/standards/rest.md#making-the-openapi-specification-available",
    "error": "Expected route /openapi/{version} to be included",
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "get open api versions",
    "passed": false,
    "received": undefined,
    "type": "requirement",
    "where": "specification",
  },
  {
    "change": {
      "location": {
        "conceptualLocation": {},
        "conceptualPath": [],
        "jsonPath": "",
        "kind": "specification",
      },
      "value": {
        "info": {
          "title": "Empty",
          "version": "0.0.0",
        },
        "openapi": "3.1.0",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/standards/rest.md#making-the-openapi-specification-available",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "list open api version",
    "passed": true,
    "received": undefined,
    "type": "requirement",
    "where": "specification",
  },
]
`;

exports[`specification rules compiled specs fails when /openapi/{versions} endpoint isn't specified 1`] = `
[
  {
    "change": {
      "location": {
        "conceptualLocation": {},
        "conceptualPath": [],
        "jsonPath": "",
        "kind": "specification",
      },
      "value": {
        "info": {
          "title": "Empty",
          "version": "0.0.0",
        },
        "openapi": "3.1.0",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/standards/rest.md#referenced-entities",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "component names",
    "passed": true,
    "received": undefined,
    "type": "requirement",
    "where": "specification",
  },
  {
    "change": {
      "location": {
        "conceptualLocation": {},
        "conceptualPath": [],
        "jsonPath": "",
        "kind": "specification",
      },
      "value": {
        "info": {
          "title": "Empty",
          "version": "0.0.0",
        },
        "openapi": "3.1.0",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/standards/rest.md#tags",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "open api version names",
    "passed": true,
    "received": undefined,
    "type": "requirement",
    "where": "specification",
  },
  {
    "change": {
      "location": {
        "conceptualLocation": {},
        "conceptualPath": [],
        "jsonPath": "",
        "kind": "specification",
      },
      "value": {
        "info": {
          "title": "Empty",
          "version": "0.0.0",
        },
        "openapi": "3.1.0",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/standards/rest.md#making-the-openapi-specification-available",
    "error": "Expected route /openapi/{version} to be included",
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "get open api versions",
    "passed": false,
    "received": undefined,
    "type": "requirement",
    "where": "specification",
  },
  {
    "change": {
      "location": {
        "conceptualLocation": {},
        "conceptualPath": [],
        "jsonPath": "",
        "kind": "specification",
      },
      "value": {
        "info": {
          "title": "Empty",
          "version": "0.0.0",
        },
        "openapi": "3.1.0",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/standards/rest.md#making-the-openapi-specification-available",
    "error": "Expected route /openapi to be included",
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "list open api version",
    "passed": false,
    "received": undefined,
    "type": "requirement",
    "where": "specification",
  },
]
`;

exports[`specification rules fails when components are not PascalCase 1`] = `
[
  {
    "change": {
      "location": {
        "conceptualLocation": {},
        "conceptualPath": [],
        "jsonPath": "",
        "kind": "specification",
      },
      "value": {
        "info": {
          "title": "Empty",
          "version": "0.0.0",
        },
        "openapi": "3.1.0",
        "x-snyk-api-stability": "wip",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/standards/rest.md#referenced-entities",
    "error": "Expected thingResourceResponse to be pascal case",
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "component names",
    "passed": false,
    "received": undefined,
    "type": "requirement",
    "where": "specification",
  },
  {
    "change": {
      "location": {
        "conceptualLocation": {},
        "conceptualPath": [],
        "jsonPath": "",
        "kind": "specification",
      },
      "value": {
        "info": {
          "title": "Empty",
          "version": "0.0.0",
        },
        "openapi": "3.1.0",
        "x-snyk-api-stability": "wip",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/standards/rest.md#tags",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "open api version names",
    "passed": true,
    "received": undefined,
    "type": "requirement",
    "where": "specification",
  },
]
`;

exports[`specification rules uncompiled specs passes when /openapi endpoint isn't specified 1`] = `
[
  {
    "change": {
      "location": {
        "conceptualLocation": {},
        "conceptualPath": [],
        "jsonPath": "",
        "kind": "specification",
      },
      "value": {
        "info": {
          "title": "Empty",
          "version": "0.0.0",
        },
        "openapi": "3.1.0",
        "x-snyk-api-stability": "wip",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/standards/rest.md#referenced-entities",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "component names",
    "passed": true,
    "received": undefined,
    "type": "requirement",
    "where": "specification",
  },
  {
    "change": {
      "location": {
        "conceptualLocation": {},
        "conceptualPath": [],
        "jsonPath": "",
        "kind": "specification",
      },
      "value": {
        "info": {
          "title": "Empty",
          "version": "0.0.0",
        },
        "openapi": "3.1.0",
        "x-snyk-api-stability": "wip",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/standards/rest.md#tags",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "open api version names",
    "passed": true,
    "received": undefined,
    "type": "requirement",
    "where": "specification",
  },
]
`;

exports[`specification rules uncompiled specs passes when /openapi/{versions} endpoint isn't specified 1`] = `
[
  {
    "change": {
      "location": {
        "conceptualLocation": {},
        "conceptualPath": [],
        "jsonPath": "",
        "kind": "specification",
      },
      "value": {
        "info": {
          "title": "Empty",
          "version": "0.0.0",
        },
        "openapi": "3.1.0",
        "x-snyk-api-stability": "wip",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/standards/rest.md#referenced-entities",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "component names",
    "passed": true,
    "received": undefined,
    "type": "requirement",
    "where": "specification",
  },
  {
    "change": {
      "location": {
        "conceptualLocation": {},
        "conceptualPath": [],
        "jsonPath": "",
        "kind": "specification",
      },
      "value": {
        "info": {
          "title": "Empty",
          "version": "0.0.0",
        },
        "openapi": "3.1.0",
        "x-snyk-api-stability": "wip",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/standards/rest.md#tags",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "open api version names",
    "passed": true,
    "received": undefined,
    "type": "requirement",
    "where": "specification",
  },
]
`;
